### PR TITLE
Use autoconfiguration to initialise components

### PR DESCRIPTION
**Pull request by rlewan**
This change makes sure that no CCD related components are being initialised if the CCD url configuration property is not provided.

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/ccd/client/CoreCaseDataApi.java [hunk 1]
```diff
@@ -67,17 +63,9 @@ CaseDetails submit(@RequestHeader(AUTHORIZATION) String authorisation,
...
-            return new JacksonDecoder(objectMapper());
-        }
-
-        @Bean
-        public ObjectMapper objectMapper() {
-            return new ObjectMapper()
-                .registerModule(new Jdk8Module())
-                .registerModule(new JavaTimeModule()).disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS)
-                .disable(SerializationFeature.FAIL_ON_EMPTY_BEANS);
+        Decoder feignDecoder(ObjectMapper objectMapper) {
```
Comment (rlewan) [inline src/main/java/uk/gov/hmcts/reform/ccd/client/CoreCaseDataApi.java]: `ObjectMapper` instance will be taken from the service which uses this client.

File: src/main/java/uk/gov/hmcts/reform/ccd/client/CoreCaseDataClientAutoConfiguration.java [hunk 2]
```diff
@@ -7,12 +7,13 @@
...
 
 @Configuration
+@ConditionalOnProperty(prefix = "core_case_data", name = "api.url")
 @EnableFeignClients(basePackages = "uk.gov.hmcts.reform.ccd.client")
 public class CoreCaseDataClientAutoConfiguration {
 
     @Bean
-    @ConditionalOnProperty(prefix = "core_case_data", name = "api.url")
-    public CoreCaseDataHealthIndicator healthIndicator(CoreCaseDataApi coreCaseDataApi) {
+    public CoreCaseDataHealthIndicator coreCaseData(CoreCaseDataApi coreCaseDataApi) {
```
Comment (rlewan) [inline src/main/java/uk/gov/hmcts/reform/ccd/client/CoreCaseDataClientAutoConfiguration.java]: This method name is used as the name displayed in the health check entry -- had to update to make it meaningful.

File: src/main/java/uk/gov/hmcts/reform/ccd/client/healthcheck/InternalHealth.java [hunk 3]
```diff
@@ -11,9 +10,9 @@
 
     @JsonCreator
     public InternalHealth(
-        @JsonProperty("status") Status status
```
Comment (rlewan) [inline src/main/java/uk/gov/hmcts/reform/ccd/client/healthcheck/InternalHealth.java]: This didn't work, not sure why. Lost an hour on it and gave up.

File: src/main/java/uk/gov/hmcts/reform/ccd/client/CoreCaseDataClientAutoConfiguration.java (hunk 2 above)
Comment (geek-soft) [reply src/main/java/uk/gov/hmcts/reform/ccd/client/CoreCaseDataClientAutoConfiguration.java]: I did not know that, good explanation. Thanks @rlewan

File: src/main/java/uk/gov/hmcts/reform/ccd/client/healthcheck/InternalHealth.java (hunk 3 above)
Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/ccd/client/healthcheck/InternalHealth.java]: There are 2 gotcha's with this class:
- It has `@JsonProperty("status")` in there already
- Does not have default no arg constructor

Here's the snippet of working (I think) type safe code:

```java
@JsonIgnoreProperties(ignoreUnknown = true)
@JsonRootName("status")
public class InternalHealth {
    @JsonUnwrapped
    private final Status status;

    @JsonCreator
    public InternalHealth(Status status) {
        this.status = status;
    }

    public Status getStatus() {
        return status;
    }
}

public class InternalHealthTest {

    private final ObjectMapper objectMapper = new ObjectMapper();

    @Before
    public void setUp() {
        objectMapper.enable(DeserializationFeature.UNWRAP_ROOT_VALUE);
    }

    @Test
    public void asd() throws IOException {
        InternalHealth health = objectMapper.readValue("{\"status\":\"UP\"}", InternalHealth.class);

        assertThat(health.getStatus()).isEqualTo(Status.UP);
    }

    @Test
    public void dsa() throws JsonProcessingException {
        String health = objectMapper.writeValueAsString(new InternalHealth(Status.DOWN));

        assertThat(health).isEqualTo("{\"status\":\"DOWN\"}");
    }
}
```

**Pull request merged**